### PR TITLE
Esirkepov to cpp

### PR DESCRIPTION
**Pull request by dpgrote**
This has been compared to the old version and gives same results to machine precision. The order 1, 2, 3, were tested on both cpu and gpu.

This is faster than the Fortran version on the gpu by a factor of ~3.

### CODE REVIEW

**Comment by ax3l** [comment]
if the test fails due to time constrains, maybe add `travis_wait 30 ...` to the long-running line.

File: Source/Particles/Deposition/CurrentDeposition.H [hunk 1]
```diff
@@ -176,4 +176,280 @@ void doDepositionShapeN(const Real * const xp, const Real * const yp, const Real
...
+// Specialized templates are defined below for orders 1, 2 and 3.
+template <int depos_order>
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+int compute_shifted_shape_factor(Real* const sx, Real x_old, int i_new) {return 0;};
+
+// Compute shape factor for order 1.
+template <>
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+int compute_shifted_shape_factor <1> (Real* const sx, Real x_old, int i_new){
+    int i = (int) x_old;
```
Comment (ax3l) [inline Source/Particles/Deposition/CurrentDeposition.H]: let's make all of those `const`:
```suggestion
    const int i = (int) x_old;
```

File: Source/Particles/Deposition/CurrentDeposition.H [hunk 2]
```diff
@@ -176,4 +176,280 @@ void doDepositionShapeN(const Real * const xp, const Real * const yp, const Real
...
+    int i_shift = i - i_new;
```
Comment (ax3l) [inline Source/Particles/Deposition/CurrentDeposition.H]:
```suggestion
    const int i_shift = i - i_new;
```

File: Source/Particles/Deposition/CurrentDeposition.H [hunk 3]
```diff
@@ -176,4 +176,280 @@ void doDepositionShapeN(const Real * const xp, const Real * const yp, const Real
...
+    Real xint = x_old - i;
```
Comment (ax3l) [inline Source/Particles/Deposition/CurrentDeposition.H]:
```suggestion
    const Real xint = x_old - i;
```

File: Source/Particles/Deposition/CurrentDeposition.H [hunk 4]
```diff
@@ -176,4 +176,280 @@ void doDepositionShapeN(const Real * const xp, const Real * const yp, const Real
...
+                         Real sdxi = 0.;
+                         for (int i=dil; i<=depos_order+1-diu; i++) {
+                             sdxi += wqx*(sx_old[i] - sx_new[i])*(sz_new[k] + 0.5*(sz_old[k] - sz_new[k]));
+                             amrex::Gpu::Atomic::Add( &Jx_arr(lo.x+i_new-1+i, lo.y+k_new-1+k, 0), sdxi);
+                         }
+                     }
+
+                     for (int k=dkl; k<=depos_order+2-dku; k++) {
+                         for (int i=dil; i<=depos_order+2-diu; i++) {
+                             Real sdyj = wq*uyp[ip]*gaminv*invvol*((sz_new[k] + 0.5*(sz_old[k] - sz_new[k]))*sx_new[i] +
```
Comment (ax3l) [inline Source/Particles/Deposition/CurrentDeposition.H]:
```suggestion
                             const Real sdyj = wq*uyp[ip]*gaminv*invvol*((sz_new[k] + 0.5*(sz_old[k] - sz_new[k]))*sx_new[i] +
```

File: Source/Particles/Deposition/CurrentDeposition.H [hunk 5]
```diff
@@ -176,4 +176,280 @@ void doDepositionShapeN(const Real * const xp, const Real * const yp, const Real
         );
 }
 
+// Compute shape factor and return index of leftmost cell where 
+// particle writes.
+// Specialized templates are defined below for orders 1, 2 and 3.
+template <int depos_order>
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+int compute_shifted_shape_factor (Real* const sx, const Real x_old, const int i_new) {return 0;};
```
Comment (ax3l) [inline Source/Particles/Deposition/CurrentDeposition.H]: the best think is probably best to keep this unimplemented, which will avoid returning invalid results for unspecialized `depos_order`s
```suggestion
int compute_shifted_shape_factor (Real* const sx, const Real x_old, const int i_new);
```

but we have to check this compiles.

Comment (ax3l) [reply Source/Particles/Deposition/CurrentDeposition.H]: I pushed a commit that updates this and fixes CI

**Comment by RemiLehe** [comment]
Looks good to me. 
I'll merge now, but will propose cosmetic changes later on, via separate PR.

**Pull request merged**